pkg/controllermanager/controller/bastion/bastion_control.go

```diff
@@ -89,6 +93,23 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 		return reconcile.Result{}, nil
 	}
 
+	// fetch associated Shoot
+	shoot := gardencorev1beta1.Shoot{}
+	shootKey := kutil.Key(bastion.Namespace, bastion.Spec.ShootRef.Name)
+	if err := r.gardenClient.Get(ctx, kutil.Key(bastion.Namespace, bastion.Spec.ShootRef.Name), &shoot); err != nil {
+		return reconcile.Result{}, fmt.Errorf("could not get shoot %v: %v", shootKey, err)
```

Comment:
if the shoot is gone, we should probably also delete the bastion, or not?

Reply:
Good point. Probably we should delete all `Bastion`s as soon as a shoot gets a deletion timestamp.

Reply:
but could there be a case that you want to troubleshoot a shoot node using a bastion where the shoot is being deleted but somehow got stuck?

Reply:
How? :) Manually? Set an owner ref? Does the graph tooling in Gardener already take care of removing nodes that have edges to a Shoot?

Reply:
> I cannot think about a scenario in which is would be required, can you?

@rfranzke me neither, just wanted to make sure that nothing is overlooked. Fine then

Reply:
> Yes, the bastion vertex as well as related edges to the shoot vertex get deleted from the graph. Not sure how it is related in this thread, though.

I was under the impression that the graph is used to perform cleanup operations in the cluster, to model the dependencies independently of Kubernetes ownerships/finalizers. There is probably some documentation, but that was just my first assumption seeing a graph in the codebase. ^^

Reply:
:) No, the graph is used to restrict which resources in the garden cluster a gardenlet is allowed to read (similar to how the `NodeAuthorizer` is implemented in Kubernetes).

Reply:
I've added code and tests to handle both missing shoots and shoots that are currently in deletion.

Reply:
However this only cleanses up the bastion on the next reconcile loop of the bastion, which could be the expiration date of the bastion - at the latest.
To clean it up faster I guess we should (rather?/in addition?) set an owner ref so that the bastion is deleted once the shoot is deleted

Reply:
Looking at the implementation of the AWS extension, this does not work. It depends on the `extensions.gardener.cloud/v1alpha1.Cluster` resource which is already gone after the `Shoot` is deleted.

So, what about the following:
1. Either prevent `Shoot` deletion entirely if there are still related `Bastion`s in the system (less preferred)
2. At the beginning of both the shoot migration and deletion flows, check if there are still related `Bastion`s in the system, and only continue if the result is `0`. In parallel, this controller here in GCM should watch `Shoot`s and requeue related `Bastion`s if the `.metadata.deletionTimestamp` of the `Shoot` was set to ensure the related `Bastion`s are being cleaned up asap (preferred).

WDYT?

Reply:
+1 for the second option

plus, we should also validate that the shoot is not in deletion when creating a new bastion. Currently we only check if the shoot exists https://github.com/gardener/gardener/pull/3974/files#diff-a56db20e6a9b4aad3584c6d1250433921d7cdf4def9f1272d3de53ba21f3dd3fR129-R139

Reply:
> In parallel, this controller here in GCM should watch Shoots and requeue related Bastions if the .metadata.deletionTimestamp of the Shoot was set to ensure the related Bastions are being cleaned up asap (preferred).

Wouldn't this be solved by the owner ref?

Reply:
No, the owner reference will make the garbage collector deleting the `Bastion` only after the `Shoot` disappeared from the store. If the shoot has only a non-`nil` `.metadata.deletionTimestamp` then nothing happens.

Reply:
> At the beginning of both the shoot migration and deletion flows, check if there are still related Bastions in the system, and only continue if the result is 0. In parallel, this controller here in GCM should watch Shoots and requeue related Bastions if the .metadata.deletionTimestamp of the Shoot was set to ensure the related Bastions are being cleaned up asap (preferred).

@rfranzke If we only trigger a reconcile for the bastions when a Shoot was deleted, what happens when a Shoot is being migrated? If I understood you correctly, the Shoot controller in the GCM would notice the existing Bastions and refuse the migration. But then nothing further happens, because the Bastion controller in the GCM does nothing. Shouldn't the Bastion controller also react to "shoots that are scheduled for migration", so that it can cleanup existing Bastions to allow the shoot migration to start? Or do we really want to just wait until all Bastions reach their natural end-of-life before migrating the shoot?

Reply:
> Shouldn't the Bastion controller also react to "shoots that are scheduled for migration", so that it can cleanup existing Bastions to allow the shoot migration to start? Or do we really want to just wait until all Bastions reach their natural end-of-life before migrating the shoot?

Yes, of course, we should! I haven't written it in my description above, sorry. We should delete all `Bastion`s if a shoot's deletion timestamp gets set or if migration is triggered (by changing its seed name).

Reply:
Looks like I've already implemented it that way ^^ https://github.com/gardener/gardener/pull/4005/files#diff-756a27d8ca9fb61d2ebea7d5b726400f0c0d4edff1965c0de26e65cd9c23c4bcR150-R166